src/sync/app.cpp

```diff
@@ -734,6 +732,113 @@ void App::link_user(std::shared_ptr<SyncUser> user,
+    {
+        refresh_access_token(sync_user, completion_block);
+    }
+    
+    void App::do_authenticated_request(Request request,
+                                       std::shared_ptr<SyncUser> sync_user,
+                                       std::function<void (Response)> completion_block) const
+    {
+        auto handler = [completion_block, request, sync_user, this](const Response& response) {
+            if (auto error = check_for_errors(response)) {
+                App::handle_auth_failure(error.value(), response, request, sync_user, completion_block);
+            } else {
+                completion_block(response);
+            }
+        };
+
+        request.headers = get_request_headers(sync_user,
+                                              request.uses_refresh_token ?
+                                              RequestTokenType::RefreshToken : RequestTokenType::AccessToken);
+        m_config.transport_generator()->send_request_to_server(request, handler);
+    }
+    
+    void App::handle_auth_failure(const AppError& error,
+                                  const Response& response,
+                                  Request request,
+                                  std::shared_ptr<SyncUser> sync_user,
+                                  std::function<void (Response)> completion_block) const
+    {
+        auto transport_generator = m_config.transport_generator();
+        auto access_token_handler = [&transport_generator,
+                                     &request,
+                                     completion_block,
+                                     response,
+                                     sync_user](const Optional<AppError>& error) {
+            if (!error) {
+                // assign the new access_token to the auth header
+                request.headers = get_request_headers(sync_user, RequestTokenType::AccessToken);
+                transport_generator->send_request_to_server(request, completion_block);
+            } else {
+                // pass the error back up the chain
+                completion_block(response);
+            }
+        };
+        
+        // Only handle auth failures
+        if (error.error_code.value() != 401) {
+            completion_block(response);
+            return;
+        }
+        
+        if (request.uses_refresh_token) {
+            if (!sync_user) {
+                sync_user->log_out();
+            }
+            completion_block(response);
+            return;
+        }
+        
+        App::refresh_access_token(sync_user, access_token_handler);
+    }
+    
+    /// MARK: - refresh access token
+    void App::refresh_access_token(std::shared_ptr<SyncUser> sync_user,
+                                   std::function<void(Optional<AppError>)> completion_block) const
+    {
+        if (!sync_user) {
+            completion_block(AppError(make_custom_error_code(ServiceErrorCode::invalid_session),
```

Comment:
can you use a `ClientErrorCode` here? Either reuse `user_not_found` or create another value that is more appropriate, but don't use the custom error category.

Reply:
Now using `ClientErrorCode`